Allow usage of block notice templates when using classic themes

nielslange opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #44538.

This PR enables merchants to display the blocks notice templates even when using classic themes. This can be accomplished by overwriting the template files or by using the following filter:

```php
add_filter( 'woocommerce_use_block_notices_in_classic_theme', '__return_true' );
```

Apart from adjusting the display logic, I've also implemented e2e tests to ensure that this option remains working. In total, we're now having the following e2e tests regarding notice templates:

- Display **default classic notice template** when using **classic theme**.
- Display **custom classic notice template on overwrite** when using **classic theme**.
- Display **default block notice template on filter** when using **classic theme**.
- Display **custom block notice template on overwrite** when using **classic theme**.
- Display **default block notice template** when using **block theme**.
- Display **default block notice template on filter** when using **block theme**.
- Display **custom block notice template on overwrite** when using **block theme**.
- Display **custom classic notice template on overwrite** when using **block theme**.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install the Storefront theme.
2. Install and activate the [storefront-child__block-notices-template.zip](https://github.com/woocommerce/woocommerce/files/14535269/storefront-child__block-notices-template.zip) theme.
3. Create a test page and add the following shortcode to it:

```
[woocommerce_cart]
```

4. Go to `/wp-admin/admin.php?page=wc-settings&tab=advanced` and set the test page of the previous step as the Cart page.
5. Go to `/wp-admin/edit.php?post_type=shop_coupon` and create a coupon.
6. Go to the frontend and add a product to the cart.
7. Go to the test page with the cart shortcode.
8. Provide the coupon code and verify that the block notice success template (green) is visible.
9. Provide the same coupon code again and verify that the block notice error template (red) is visible.
10. Delete the product from the cart and verify that the block notice info template (blue) is visible.
11. Activate the Storefront theme.
12. Install and activate the [Code Snippets](https://wordpress.org/plugins/code-snippets/) plugin.
13. Add the following code snippet:

```php
add_filter( 'woocommerce_use_block_notices_in_classic_theme', '__return_true' );
```

14. Repeat steps 6. until 10. and verify that the block notice templates remain visible.

Apart from the previous testing instructions, also have a look at the added and adjusted e2e tests and verify that all e2e tests are passing.

### Screenshots

<table>
<tr>
<td valign="top">Block notice success template:
<br><br>
<img width="866" alt="Screenshot 2024-02-27 at 14 05 11" src="https://github.com/woocommerce/woocommerce/assets/3323310/dbae242f-4359-462d-996f-ec5c007e9b98">
</td>
<td valign="top">Block notice error template:
<br><br>
<img width="862" alt="Screenshot 2024-02-27 at 14 05 29" src="https://github.com/woocommerce/woocommerce/assets/3323310/d3297944-8e51-4b95-94ef-099b2b7de432">
</td>
<td valign="top">Block notice info template:
<br><br>
<img width="858" alt="Screenshot 2024-02-27 at 14 05 43" src="https://github.com/woocommerce/woocommerce/assets/3323310/91a5710d-3215-4228-a448-889b264e98e4">
</td>
</tr>
</table>

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Allow usage of block notice templates when using classic themes.

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

nielslange committed:
Allow block notice templates in classic themes

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

nielslange committed:
Fix e2e constants file

opr commented on plugins/woocommerce/src/Blocks/Domain/Services/Notices.php:
```diff
@@ -42,11 +42,11 @@ public function __construct( Package $package ) {
 		}
 
+		add_filter( 'woocommerce_kses_notice_allowed_tags', [ $this, 'add_kses_notice_allowed_tags' ] );
+		add_action( 'wp_head', [ $this, 'enqueue_notice_styles' ] );
```
With this change, are notice styles going to be enqueued on every page load for classic themes too? Is this a problem?

nielslange replied:
Thanks for mentioning this, @opr. Enqueueing the block notice styles for classic themes is necessary in case a merchant, that is using a classic theme, is using block notices. Without enqueueing the block notice styles, the notice would look like this:

<img width="840" alt="Screenshot 2024-03-05 at 15 32 36" src="https://github.com/woocommerce/woocommerce/assets/3323310/8a338d61-442a-4e3a-94df-a6487885b472">

The block notice styles, that we're enqueueing, add 2.7 kB to the page load. This impact should not be noticeable to merchants or sellers. 

<img width="2560" alt="Screenshot 2024-03-05 at 15 36 04" src="https://github.com/woocommerce/woocommerce/assets/3323310/1ec70819-425d-46bd-8249-19890b663286">

opr replied:
> The block notice styles, that we're enqueueing, add 2.7 kB to the page load. This impact should not be noticeable to merchants or sellers.

Sure, but if we can avoid this somehow then it would be better. Can we think of a way of allowing classic themes to signal that they want to use block notices, a filter for example?

Before this PR, the notices were only enqueued wen `is_block_theme` is true, however imagine if we had something like

`apply_filters( 'woocommerce_blocks_use_block_notice_styles', is_block_theme() )`

What are your thoughts on that?

nielslange replied:
As suggested, I've added the filter, but slightly adjusted the name, as discussed in p1709635632195569-slack-C8X6Q7XQU:

```php
apply_filters( 'woocommerce_use_block_notices_in_classic_theme', false ) 
```

opr commented on plugins/woocommerce-blocks/tests/e2e/tests/cart/cart-checkout-block-notices.shopper.classic_theme.spec.ts:
```diff
@@ -60,8 +61,7 @@ test.describe( 'Shopper → Classic Notice Templates', () => {
 
-		// We're explicitly checking for the following CSS classes here:
-		// .woocommerce-notices-wrapper .woocommerce-message
+		// We're explicitly checking the CSS classes.
```
I'm not sure this comment (and the other instances like it) are adding much value, giving the reason why we are checking them would be more useful.

nielslange replied:
Great catch, @opr. I've changed the comments according to their corresponding type.

opr reviewed (changes requested):
Works OK for me, but left a couple of comments that need answers/addressing before approving. Thanks Niels

nielslange committed:
Improve inline comments

nielslange commented:
Thanks for your review, @opr. I've addressed your feedback. Currently, there are two test failing. Nadir firstly mentioned these failing tests in p1709562740110129-slack-C03CPM3UXDJ and Bart mentioned that he'll be looking into them today. As these failing tests seem unrelated to this PR, you could go ahead and re-review the PR again, when you have time.

nielslange committed:
Add filter and enhance e2e tests

nielslange committed:
Adjust test names

nielslange commented on plugins/woocommerce-blocks/tests/e2e/bin/themes/storefront-child__block-notices-filter/functions.php:
```diff
@@ -0,0 +1,3 @@
+<?php
+
+add_filter( 'woocommerce_use_block_notices_in_classic_theme', '__return_true' );
```
This filter is used to verify that classic theme developers can trigger the block notices via filter.

nielslange commented on plugins/woocommerce-blocks/tests/e2e/bin/themes/twentytwentyfour-child__block-notices-filter/functions.php:
```diff
@@ -0,0 +1,3 @@
+<?php
+
+add_filter( 'woocommerce_use_block_notices_in_classic_theme', '__return_false' );
```
This filter is used to verify that block theme developers cannot trigger the classic notices via filter.

opr reviewed (dismissed):
This looks much better now Niels - there are some checks failing - https://github.com/woocommerce/woocommerce/actions/runs/8200519026/job/22427557137?pr=45164

Please could you address those and get a fresh review from someone else on Rubik as I'll be AFK for a week now.

nielslange committed:
Adress PHP lint errors

nielslange committed:
Adress PHP lint errors

nielslange committed:
Adress PHP lint errors

wavvves reviewed (approved):
Wow, I bet this started out as something that would be tiny, and grew into a lot of changes and tests 😅

Nice work @nielslange, approving 👍🏼

nielslange commented:
Thanks for reviewing this PR, @wavvves. This PR indeed started small and then kept growing and growing. You can probably tell by the amount of test-related files. 😆 That said, thanks again for reviewing it. 🙌

senadir reviewed (commented):
FWIW I just ended up testing this and things works fine! great work with the tests Niels.

PaulSchiretz commented:
Very nice fix! Just one thing, if i do NOT opt in to show block notices in a classic theme wp_enqueue_style( 'wc-blocks-style' ); is still called on wp_head and there is no way to dequeue it.

```
	/**
	 * Initialize notice hooks.
	 */
	public function init() {
		add_action(
			'after_setup_theme',
			function() {
				/**
				 * Allow classic theme developers to opt-in to using block notices.
				 *
				 * @since 8.8.0
				 * @param bool $use_block_notices_in_classic_theme Whether to use block notices in classic theme.
				 * @return bool
				 */
				if ( wp_is_block_theme() || apply_filters( 'woocommerce_use_block_notices_in_classic_theme', false ) ) {
					add_filter( 'wc_get_template', [ $this, 'get_notices_template' ], 10, 5 );
				}
			}
		);

		add_filter( 'woocommerce_kses_notice_allowed_tags', [ $this, 'add_kses_notice_allowed_tags' ] );
		add_action( 'wp_head', [ $this, 'enqueue_notice_styles' ] );
	}
```

Wouldn't it make more sense to also check the conditionals in enqueue_notice_styles?
The change would be very minor but would allow us to get rid of some unused css.

Simply replace:
```
	/**
	 * Replaces all notices with the new block based notices.
	 *
	 * @return void
	 */
	public function enqueue_notice_styles() {
		wp_enqueue_style( 'wc-blocks-style' );
	}
```

with:

```
	/**
	 * Replaces all notices with the new block based notices.
	 *
	 * @return void
	 */
	public function enqueue_notice_styles() {
		if ( wp_is_block_theme() || apply_filters( 'woocommerce_use_block_notices_in_classic_theme', false ) ) {
			wp_enqueue_style( 'wc-blocks-style' );
		}
	}
```

I know I'm late to the party, just hope someone reads this ;-)

nielslange commented:
Thanks for reaching out, @PaulSchiretz. Initially, I kept the call `add_action( 'wp_head', [ $this, 'enqueue_notice_styles' ] );` outside the conditional logic in case a merchant, that is using a classic theme, is overwriting the notice templates and wants to use the block notice template templates. Later, we've introduced the `woocommerce_use_block_notices_in_classic_theme` filter. With that filter in mind, we could indeed move `add_action( 'wp_head', [ $this, 'enqueue_notice_styles' ] );` into the conditional logic.

Instead of

```php
/**
 * Replaces all notices with the new block based notices.
 *
 * @return void
 */
public function enqueue_notice_styles() {
	if ( wp_is_block_theme() || apply_filters( 'woocommerce_use_block_notices_in_classic_theme', false ) ) {
		wp_enqueue_style( 'wc-blocks-style' );
	}
}
```

we can also move the action into the conditional statement

```php

/**
 * Initialize notice hooks.
 */
public function init() {
	add_action(
		'after_setup_theme',
		function () {
			/**
			 * Allow classic theme developers to opt-in to using block notices.
			 *
			 * @since 8.8.0
			 * @param bool $use_block_notices_in_classic_theme Whether to use block notices in classic theme.
			 * @return bool
			 */
			if ( wp_is_block_theme() || apply_filters( 'woocommerce_use_block_notices_in_classic_theme', false ) ) {
				add_filter( 'wc_get_template', [ $this, 'get_notices_template' ], 10, 5 );
				add_action( 'wp_head', [ $this, 'enqueue_notice_styles' ] );
			}
		}
	);

	add_filter( 'woocommerce_kses_notice_allowed_tags', [ $this, 'add_kses_notice_allowed_tags' ] );
}
```

However, this would mean that all merchants, who are overwriting the classic notice templates with the block notice templates, would also have to add the following call to their site, so that the required styles are loaded:

```php
add_filter( 'woocommerce_use_block_notices_in_classic_theme', '__return_true' );
```

Given that the corresponding styles file is only 2.7kb big, we've decided against this.

PaulSchiretz commented:
@nielslange I see, makes total sense. Thanks for clarifiying this to me! I kind of went the other route, once the block templates where the default i tweaked my theme to incooperate the block templates then when the change was reversed i tweaked the classic templates to look like the block templates. Now that this is supported via hook i was wondering if i missed some spot once i saw the additional css file. Eather way as you also said, better to leave the css in than cause another confusion :-) Thanks for the quick clarification!

nielslange commented:
> Thanks for the quick clarification!

Thank you for reaching out to clarify this, @PaulSchiretz. As for overwriting the notice templates, I posted [Tutorial: Overriding notice templates](https://developer.woocommerce.com/2024/02/16/tutorial-overriding-notice-templates/) a few weeks ago. You can either use the hook or manually overwrite the notice template files. 

That said, thanks again for reaching out, and please don't't hesitate to contact us if you have more questions or suggestions. 🙌

tiago-123 commented:
> ```
> add_filter( 'woocommerce_use_block_notices_in_classic_theme', '__return_true' );
> ```

It works like a charm! Thanks @nielslange!

nielslange commented:
> It works like a charm! Thanks @nielslange!

Yeah! You're very welcome, @tiago-123. 🙌
